sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,326 @@
+after the device link and post link steps.
+
+Regardless of the AOT target, the flow is similar, only modifying the offline
+compiler that is used to create the target device image.  It is expected that
+the offline compiler will also use unique command lines specific to the tool
+to create the image.
+
+To support the needed option passing triggered by use of the
+`-Xsycl-target-backend` option and implied options based on the optional
+device behaviors for AOT compilations for GPU new command line interfaces
+are needed to pass along this information.
+
+| Target | Triple        | Offline Tool   | Option for Additional Args |
+|--------|---------------|----------------|----------------------------|
+| CPU    | spir64_x86_64 | opencl-aot     | `--cpu-tool-arg=<arg>`     |
+| GPU    | spir64_gen    | ocloc          | `--gen-tool-arg=<arg>`     |
+| FPGA   | spir64_fpga   | aoc/opencl-aot | `--fpga-tool-arg=<arg>`    |
+
+*Table: Ahead of Time Info*
+
+To complete the support needed for the various targets using the
+`clang-linker-wrapper` as the main interface, a few additional options will
+be needed to communicate from the driver to the tool.  Further details of usage
+are given further below.
+
+| Option Name                  | Purpose                                      |
+|------------------------------|----------------------------------------------|
+| `--fpga-tool-deps=<arg>`     | Comma separated list of dependency files used for FPGA hardware compiles using `aoc` |
+| `--parallel-link-sycl=<arg>` | Provide the number of parallel jobs that will be used when processing with `llvm-foreach` |
+| `--no-sycl-device-lib=<arg>` | Provide the list of device libraries to restrict from linking during device link |
+
+*Table: Additional Options for clang-linker-wrapper*
+
+The `clang-linker-wrapper` provides an existing option named `-wrapper-jobs`
+that may be useful for our usage instead of creating a new option specific
+to `llvm-foreach` processing.
+
+#### spir64_gen support
+
+Compilation behaviors involving AOT for GPU involve an additional call to
+the OpenCL Offline compiler (OCLOC).  This call occurs after the post-link
+step performed by sycl-post-link.  Additional options passed by the user
+via the `-Xsycl-target-backend=spir64_gen <opts>` command as well as the
+implied options set via target options such as `-fsycl-targets=intel_gpu_skl`
+will be processed by a new options to the wrapper, `--gen-tool-arg=<arg>`
+
+#### spir64_fpga support
+
+Compilation behaviors involving AOT for FPGA involve an additional call to
+the either `aoc` (for Hardware) or `opencl-aot` (for Simulation).  This call
+occurs after the post-link step performed by sycl-post-link.  Additional
+options passed by the user via the `-Xsycl-target-backend=spir64_fpga <opts>`
+command will be processed by a new options to the wrapper,
+`--fpga-tool-arg=<arg>`
+
+The FPGA target also has support for additional generated binaries that
```

Comment:
I just want to clarify something here. With this new model, it sounds like it's possible to package binaries like aoco, aocr, or aocx into the fat object files that the link stage can pick up, correct? If yes, then I'm not sure if Diagram 1 is exactly right, since it sounds like you may have an "Offline Compile" stage there. Unless you're implying that there could be a AOT compile stage as part of the "Device Compile".

Reply:
The case of using aocx/aoco/aocr files in object files/archives is not explicitly called out and the diagram is more a reflection of the general AOT compilation case.  The usage of the FPGA specific device types should be spelled out more succinctly.